### DISCUSSION TITLE
Why do we have def mainClass: Target[Option[String]] instead of Target[String]?

### DISCUSSION THREAD

**Discussion by jk-1** [com-lihaoyi/mill#2074] [Q&A] [answered]
For what purpose is 
`def mainClass: Target[Option[String]] `
instead of not just 
`def mainClass: Target[String]`

I thought that defining

```
object bar extends ScalaModule with ScalafmtModule {
  def scalaVersion = "2.13.1" // scalaVer
  override def mainClass: Target[Option[String]] = T { None } // No main class implemented
}

```
would cause `mill show _.run` command just skip the `mill show bar.run` and go on running later modules but in this case it failed during `bar.run` and no later modules were compiled or run as you see from the output:

```
jk@t12:~/workspace/mill-exp$ mill show _.run
Compiling /home/jk/workspace/mill-exp/build.sc
...
[1/1] show [1/1] show > [239/420] foo.compile 
[info] compiling 3 Scala sources and 1 Java source to /home/jk/workspace/mill-exp/out/foo/compile/dest/classes ...
[info] done compiling
[1/1] show > [266/420] bar.finalMainClass 
1 targets failed
show 1 targets failed
bar.finalMainClass No main class specified or found
jk@t12:~/workspace/mill-exp$ echo $?
1
```
It also seems that if target has main method and I still use

`  override def mainClass: Target[Option[String]] = T { None } `

and run `mill show _.run` then mill also runs that target with `mainClass = None` setting.

In what case should we then use 

`  override def mainClass: Target[Option[String]] = T { None } // No main class implemented`

How can I implement command like: Run all modules with main and gracefully skip modules without main?

My versions are:
```
Mill Build Tool version 0.9.8
Java version: 11.0.16, vendor: Ubuntu, runtime: /usr/lib/jvm/java-11-openjdk-amd64
Default locale: en_US, platform encoding: UTF-8
OS name: "Linux", version: 5.4.0-131-generic, arch: amd64
```

**Comment by lefou** [comment, marked as answer]
I find Mills answer quite clear:

```bash
$ mill inspect main.mainClass
[1/1] inspect 
main.mainClass(JavaModule.scala:54)
    Allows you to specify an explicit main class to use for the `run` command.
    If none is specified, the classpath is searched for an appropriate main
    class to use if one exists

Inputs:

```

Looks like you can't skip the run command by just providing no main class.

> How can I implement command like: Run all modules with main and gracefully skip modules without main?

That's currently not possible out of the box.

You could use `mill --keep-going _.run` and ignore the failed targets, or write your own command to skip when needed.

**Reply by jk-1** [reply]
Thank's for the quick response and pointing the `mill inspect` docs. `--keep-going` might help here. Or perhaps rewrite run so that `mainClass = None` would skip it, else do `super.run`. And not having the `def mainClass` at all for default case `None` functionality (=the classpath is searched for an appropriate main class to use if one exists).

**Reply by jk-1** [reply]
Should we have a new feature in mill:

A way to create some kind of sequence or iterator of all Tasks in a build.sc file.

User could then filter that e.g. based on type and e.g. sort or execute subset of tasks. Like compile all Java tasks, test all scala tasks or run all tasks having main -function.

**Comment by nafg** [comment]
You could also define a top -level task to call run on the modules you want, or use {} syntax to tell the CLI which modules. Opt-in instead of opt-out, if that's an option.
